improve-ephemeral-cluster-handling.md

```diff
@@ -0,0 +1,419 @@
+  value of the `ephemeral` label matches the value of the `app` option.
+
+* If labels and options match, an ephemeral cluster will always be deleted if
+  the replica count of the driver deployment is 0. This means that the deployment
+  has been scaled to 0 either as part of a deletion, redeploy, or a scale operation.
+  It means that a driver that is part of the current deployment will not come back,
+  or at least will not come back for some indeterminate amount of time.  If it does,
+  it will make another cluster.
+
+* If labels and options match, an ephemeral cluster will always be deleted if
+  the driver replicationcontroller cannot be found. This means that the
+  deployment has been deleted and is never coming back (note that a *different*
+  deployment spawned by the same deploymentconfig *could* be created, but it's
+  a different deployment so doesn't get to use this cluster).
+
+*  If labels and options match and the replica count of the driver deployment
+   is 1, an ephemeral cluster will be deleted if the value of `app-status` is
+   `completed`. Since the application completed and the driver has not been scaled up
+   (that is, there is no other driver pod using the cluster) the cluster serves
+   no further purpose. The expectation is that application has run its course and
+   will not be restarted.
+
+   (In reality, because of the limitation of restart semantics for deploymentconfigs,
+   the driver application *will* actually be restarted after a time if an operator
+   does not delete the driver deploymentconfig or scale it to 0. However, since these
+   things may happen after the driver pod has exited, the cluster must be deleted
+   since the driver pod itself is the means of deletion. If a cluster is left and
+   the driver deploymentconfig is deleted while in CrashLoopBackoff, the cluster will
+   be leaked)
+
+* Ephemeral clusters will always be left if the driver replica count is > 1.
+  We don't yet have a use case for scaling a driver, but we can't stop it, and
+  if multiple driver pods are using the same cluster then we can't allow one
+  of them to delete the cluster while others may still be using it.
+
+* An ephemeral cluster will be left if the replica count is 1 and the `app-status`
+  is `terminated`. In this case we expect the driver to be restarted. The
+  application was stopped early, and the most likely case is that a driver pod
+  was rescheduled or killed purposely by an operator to case a restart. In this
+  scenario it's much quicker to allow a new driver pod to use the same cluster.
+
+  If the intention was to halt the application entirely, then the driver deploymentconfig
+  would have been deleted or the deployment would have been scaled to 0.
+
+### Application startup semantics in `start.sh`
+
+There are a few rules governing startup of a driver pod:
+
+* If a driver pod has a cluster name specified in the environment, it just
```

Comment:
probably should use future tense.  After all, the spec was written before the code, right? :)